compiler/compiler.go

```diff
@@ -86,45 +86,180 @@ func compileFileDescriptorSets(files, importPaths []string, includeImports bool)
+		return err
+	}
+	for _, md := range fd.GetMessageType() {
+		if err := resolveMessageOptions(r, md); err != nil {
+			return err
+		}
+	}
+	for _, ed := range fd.GetEnumType() {
+		if err := resolveEnumOptions(r, ed); err != nil {
+			return err
+		}
+	}
+	for _, sd := range fd.GetService() {
+		if err := resolveServiceOptions(r, sd); err != nil {
+			return err
+		}
+	}
+	for _, fd := range fd.GetExtension() {
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resolveMessageOptions(r resolver, md *pb.DescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, md.GetOptions()); err != nil {
+		return err
+	}
+	for _, fd := range md.GetField() {
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, fd := range md.GetExtension() {
+		if err := resolveUninterpretedOptions(r, fd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, nestedMD := range md.GetNestedType() {
+		if err := resolveMessageOptions(r, nestedMD); err != nil {
+			return err
+		}
+	}
+	for _, ed := range md.GetEnumType() {
+		if err := resolveEnumOptions(r, ed); err != nil {
+			return err
+		}
+	}
+	for _, erd := range md.GetExtensionRange() {
+		if err := resolveUninterpretedOptions(r, erd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	for _, od := range md.GetOneofDecl() {
+		if err := resolveUninterpretedOptions(r, od.GetOptions()); err != nil {
+			return err
 		}
 	}
 	return nil
 }
 
-func fullName(names []*pb.UninterpretedOption_NamePart) protoreflect.FullName {
-	// TODO: extend for field references and extensions of extensions.
-	str := protoreflect.FullName((*names[0].NamePart)[1:])
-	return str
+func resolveEnumOptions(r resolver, ed *pb.EnumDescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, ed.GetOptions()); err != nil {
+		return err
+	}
+	for _, evd := range ed.GetValue() {
+		if err := resolveUninterpretedOptions(r, evd.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func resolveServiceOptions(r resolver, sd *pb.ServiceDescriptorProto) error {
+	if err := resolveUninterpretedOptions(r, sd.GetOptions()); err != nil {
+		return err
+	}
+	for _, md := range sd.GetMethod() {
+		if err := resolveUninterpretedOptions(r, md.GetOptions()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type messageWithOptions interface {
+	proto.Message
+	GetUninterpretedOption() []*pb.UninterpretedOption
+}
+
+func resolveUninterpretedOptions(r resolver, opts messageWithOptions) error {
+	if opts == nil || reflect.ValueOf(opts).IsNil() {
+		return nil
+	}
+
+	for _, opt := range opts.GetUninterpretedOption() {
+		protoText := UninterpretedOptionAsProtoText(opt.Name, opt)
+		m := opts.ProtoReflect().New().Interface()
+		o := prototext.UnmarshalOptions{Resolver: r}
+		if err := o.Unmarshal([]byte(protoText), m); err != nil {
```

Comment:
I don't live round-tripping through prototext, is there a better way?

Reply:
As discussed, there probably is but that is likely going to require larger surgery. The current code already round-trips through prototext - this PR adds a little more of that for basic values on known option fields, but that is rather trivial in comparison.

With protoreflect, it should be possible to make this work without prototext, and in that case we could also get rid of protog/registry - it's only purpose is for the prototext unmarshaler.

I'll see how easy it is to hook into the earlier parts of the compiler - at the moment, resolving options is a pass after the FDS is created and we can't plumb through arbitrary data structures through that. UninterpretedOption seemed tailor made for it, but it does mean going via a string and not a data structure.

Reply:
It was too much to avoid the round-tripping through prototext, so I still do it, but in a more limited fashion. All the option names (fields and extensions) are looked-up with protoreflect. The prototext unmarshaler is only used for values that are aggregate options.